### DISCUSSION TITLE
Wrapping an image around a 3D object

### DISCUSSION THREAD

**Discussion by mirko-horvacki** [ImageMagick/ImageMagick#6915] [Help]
### ImageMagick version

7.1.1.9

### Operating system, version and so on

Windows

### Description

Hi,

I am using ImageMagick to wrap artwork around 3D object.

This is the artwork that is being wrapped:
![MO6676_POS1_position-preview](https://github.com/ImageMagick/ImageMagick/assets/23114355/f896a2b0-b174-49a5-88d4-4d505629edab)

And this is the vector map being used:
[MO6676_POS1_vectormap.zip](https://github.com/ImageMagick/ImageMagick/files/13546126/MO6676_POS1_vectormap.zip)

Product image is:
![MO6676_POS1_product](https://github.com/ImageMagick/ImageMagick/assets/23114355/17a41d82-e00a-4dbb-8817-e3fae38178db)

For some reason - result almost in all cases have black lines on a places where border between 3d object and product image is. 

Here is how it looks:
![MO6676_1](https://github.com/ImageMagick/ImageMagick/assets/23114355/fe44594b-2f8c-4fdc-acd2-cc06cb1731d9)

Command that is being used is:
`var command = @$"{positionPreviewImage} -resize ""{productImg.Width}x{productImg.Height}^!"" {vectorMapImage} -resize 400%% -virtual-pixel Edge -compose Distort -define -compose:args=100%%x100%% -composite -resize 25%% {productImage} -compose Overlay -composite {result}"`

C# is used to invoke command and pass variables.

Here is one result without using super sampling:
`var command = @$"{positionPreviewImage} -resize ""{productImg.Width}x{productImg.Height}^!"" {vectorMapImage} -virtual-pixel Edge -compose Distort -define -compose:args=100%%x100%% -composite {productImage} -compose Overlay -composite {result}"`
![MO6676_1](https://github.com/ImageMagick/ImageMagick/assets/23114355/88762cfa-ff1e-4071-b027-ac58cad8f1f8)

Do you maybe know what I am doing wrong? 

Kind regards,

Mirko

**Comment by snibgo** [comment]
I can't download your product image, perhaps because the URL has "private-user".

`-define -compose:args=100%%x100%%` is bad. It should not have a hyphen (dash) before `compose:`.

The problem with the black lines is because your `MO6676_POS1_vectormap.tif` is bad. It has anti-aliasing, which creates the problem. Look closely at the pixels near the bottom, at the green/black boundary. Pixels there are a blend of green and black, which gives a displacement that is wrong.

There are similar problems around each bottle shape, but these are not so obvious.

**Comment by mirko-horvacki** [comment]
Hi,

thanks for fast response!

I tried with new vector map, which is not using AA:
[MO6676_POS1_vectormap .zip](https://github.com/ImageMagick/ImageMagick/files/13547120/MO6676_POS1_vectormap.zip)

And result still have black lines:
![MO6676_1](https://github.com/ImageMagick/ImageMagick/assets/23114355/68d669b7-abcf-4989-aa24-50b43560dea0)

I also updated command:
`var command = @$"{positionPreviewImage} -resize ""{productImg.Width}x{productImg.Height}^!"" {vectorMapImage} -resize 400%% -virtual-pixel Edge -compose Distort -define compose:args=100%%x100%% -composite -resize 25%% {productImage} -compose Overlay -composite {result}";`

Here is product image:
![MO6676_POS1_product](https://github.com/ImageMagick/ImageMagick/assets/23114355/dcec89fc-a593-490b-9c51-e5a983f30b9c)

**Comment by snibgo** [comment]
I still can't download the product image. Please zip it and upload that.

The vectormap is improved, but still has isolated antialiasing. See this enlargement:

![v](https://github.com/ImageMagick/ImageMagick/assets/33812368/c35fa81f-25a6-41d3-8433-f6f0310d6050)

Your command includes:
```
{vectorMapImage} -resize 400%%
```
So you are resizing the vectormap, and this introduces antialiasing, a mixture of green and black. You can see this by making debugging images, which you can examine.

```
{vectorMapImage} -resize 400%% +write x.png
```

**Comment by mirko-horvacki** [comment]
Hm,

tried now without that resize:
`var command = @$"{positionPreviewImage} -resize ""{productImg.Width}x{productImg.Height}^!"" {vectorMapImage} -virtual-pixel Edge -compose Distort -define compose:args=100%%x100%% -composite {productImage} -compose Overlay -composite {result}";`

With same vector map that has that small AA parts.

Result has lines:
![MO6676_1](https://github.com/ImageMagick/ImageMagick/assets/23114355/23bab731-5b24-4620-883d-e15aa07d8bdf)

Here is zipped product image:
[MO6676_POS1_product.zip](https://github.com/ImageMagick/ImageMagick/files/13547622/MO6676_POS1_product.zip)

I will try tomorrow with vector map that has no AA at all.

**Comment by snibgo** [comment]
With your latest vectorMapImage, where `haier.png` is your text image, and this Windows BAT command...
```
%IMG7%magick ^
  haier.png ^
  -resize "1000x874^!" ^
  MO6676_POS1_vectormap.tif ^
  -virtual-pixel Edge -compose Distort -define compose:args=100%%x100%% -composite ^
  MO6676_POS1_product.jpg ^
  -compose Overlay -composite ^
  x.jpg
```
... I get this result, with no black lines:

![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/e5e005cc-7eab-4754-b0a7-d7dcb5c30171)

The jaggies around the letters can be reduced with supersampling. Make `MO6676_POS1_vectormap.tif` at four times the linear size. I think there is also a bit-depth problem, and that file should be 32-bit/channel/pixel floating-point.

`MO6676_POS1_vectormap.tif` has a bigger problem: adjacent pixels have the same values, so the gradation isn't smooth, so multiple output pixels are taken from the same locations. This is the biggest cause of jaggies. How did you create that file?

**Comment by mirko-horvacki** [comment]
Hi,
thanks a loot for help!

Last week we were trying to create 32bit maps that would work, but without success. But AA was cause of black lines, you were right.
